Txpool pending and queued sets are never pruned, causing eventual overflow or DOS

## Txpool pending and queued sets are never pruned, causing eventual overflow or DOS

### Description

There is no mechanism to reduce transactions that might be eventually valid from the pending or queued sets, causing perpetual growth. 

### Your environment

* OS and version Ubuntu 20
* version of the Polygon SDK 6efebc6a92f885082620ed2a8d59e37d4d27074c
* branch that causes this issue `develop`

### Steps to reproduce

* Tell us how to reproduce this issue <br />
Create a chain with 4 validators. Create an a premine with a balance of 100 ETH. Submit txns with nonce 0...1000 and submit **them immediately without waiting for receipt**, each sending 1 ETH to some random account. Observe the first ~100 transactions get mined. Now the txn pool `pendingQueue` should contain 900 transactions.

These 900 transactions will constantly be popped then re-pushed back onto the queue because they are thrown with a recoverable error which pushes them back onto the queue

https://github.com/0xPolygon/polygon-sdk/blob/6efebc6a92f885082620ed2a8d59e37d4d27074c/consensus/ibft/ibft.go#L406

```
2021-11-18T16:24:28.038-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
```

* Where the issue is, if you know <br />

* Which commands triggered the issue, if any

`./main server`

### Expected behaviour

Transactions that are invalid for a long time should eventually be dropped. 

### Actual behaviour

Transactions that are invalid are constantly re-pushed onto the queue as most of the errors thrown here are recoverable

https://github.com/0xPolygon/polygon-sdk/blob/6efebc6a92f885082620ed2a8d59e37d4d27074c/state/executor.go#L358

### Logs

`txpool_status` on our testnet validator nodes

```
{ jsonrpc: '2.0', id: null, result: { pending: 0, queued: 1 } }
{ jsonrpc: '2.0', id: null, result: { pending: 0, queued: 42 } }
{ jsonrpc: '2.0', id: null, result: { pending: 0, queued: 329 } }
{ jsonrpc: '2.0', id: null, result: { pending: 3292, queued: 1561 } }
{ jsonrpc: '2.0', id: null, result: { pending: 3968, queued: 22 } }
{ jsonrpc: '2.0', id: null, result: { pending: 1145, queued: 14379 } }
{ jsonrpc: '2.0', id: null, result: { pending: 69, queued: 5711 } }
{ jsonrpc: '2.0', id: null, result: { pending: 109, queued: 5629 } }
{ jsonrpc: '2.0', id: null, result: { pending: 256, queued: 5559 } }
{ jsonrpc: '2.0', id: null, result: { pending: 0, queued: 42 } }
{ jsonrpc: '2.0', id: null, result: { pending: 60, queued: 5736 } }
{ jsonrpc: '2.0', id: null, result: { pending: 3706, queued: 1641 } }
{ jsonrpc: '2.0', id: null, result: { pending: 3630, queued: 1389 } }
{ jsonrpc: '2.0', id: null, result: { pending: 0, queued: 204 } }
```

```
2021-11-18T16:23:40.026-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:23:48.026-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:23:56.022-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:04.025-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:12.027-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:20.027-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:20.027-0500 [INFO]  polygon.consensus.ibft: picked out txns from pool: num=0 remaining=985
2021-11-18T16:24:28.038-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:28.038-0500 [INFO]  polygon.consensus.ibft: picked out txns from pool: num=0 remaining=985
2021-11-18T16:24:36.024-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:36.024-0500 [INFO]  polygon.consensus.ibft: picked out txns from pool: num=0 remaining=985
2021-11-18T16:24:44.028-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:52.026-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:25:00.025-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:25:00.025-0500 [INFO]  polygon.consensus.ibft: picked out txns from pool: num=0 remaining=985
2021-11-18T16:25:08.024-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:25:08.024-0500 [INFO]  polygon.consensus.ibft: picked out txns from pool: num=0 remaining=985
2021-11-18T16:25:16.023-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
```

Please paste any logs here that demonstrate the issue, if they exist

### Proposed solution

There should be some sort of periodic pool pruning as in geth. Geth has a periodic mechanism:

https://github.com/maticnetwork/bor/blob/a9c71eff578c652ddfe00a73267454bed5957ed0/core/tx_pool.go#L1137

And the algorithms:

https://github.com/maticnetwork/bor/blob/a9c71eff578c652ddfe00a73267454bed5957ed0/core/tx_pool.go#L1377
https://github.com/maticnetwork/bor/blob/a9c71eff578c652ddfe00a73267454bed5957ed0/core/tx_pool.go#L1462

--- a/txpool/account.go
+++ b/txpool/account.go
@@ -150,6 +150,7 @@ type account struct {
 	init               sync.Once
 	enqueued, promoted *accountQueue
 	nextNonce          uint64
+	demotions          uint
 }
 
 // getNonce returns the next expected nonce for this account.

--- a/txpool/txpool.go
+++ b/txpool/txpool.go
@@ -321,7 +321,7 @@ func (p *TxPool) Pop(tx *types.Transaction) {
 	// pop the top most promoted tx
 	account.promoted.pop()
 
-	//	successfully popping an account reset its demotions count to 0
+	//	successfully popping an account resets its demotions count to 0
 	account.demotions = 0
 
 	// update state

--- a/txpool/txpool.go
+++ b/txpool/txpool.go
@@ -377,9 +377,6 @@ func (p *TxPool) Drop(tx *types.Transaction) {
 	dropped = account.enqueued.clear()
 	clearAccountQueue(dropped)
 
-	//	reset the demotions counter
-	account.demotions = 0
-
 	p.eventManager.signalEvent(proto.EventType_DROPPED, tx.Hash)
 	p.logger.Debug("dropped account txs",
 		"num", droppedCount,
@@ -400,6 +397,9 @@ func (p *TxPool) Demote(tx *types.Transaction) {
 		)
 
 		p.Drop(tx)
+
+		//	reset the demotions counter
+		account.demotions = 0
 
 		return
 	}

--- a/txpool/txpool.go
+++ b/txpool/txpool.go
@@ -388,9 +388,17 @@ func (p *TxPool) Drop(tx *types.Transaction) {
 	)
 }
 
+//	Demote excludes an account from being further processed during block building
+//	due to a recoverable error. If an account has been demoted too many times (maxAccountDemotions),
+//	it is Dropped instead.
 func (p *TxPool) Demote(tx *types.Transaction) {
 	account := p.accounts.get(tx.From)
 	if account.demotions == maxAccountDemotions {
+		p.logger.Debug(
+			"Demote: threshold reached - dropping account",
+			"addr", tx.From.String(),
+		)
+
 		p.Drop(tx)
 
 		return

--- a/txpool/txpool.go
+++ b/txpool/txpool.go
@@ -21,6 +21,10 @@ const (
 	txSlotSize  = 32 * 1024  // 32kB
 	txMaxSize   = 128 * 1024 //128Kb
 	topicNameV1 = "txpool/0.1"
+
+	//	maximum allowed number of times an account
+	//	was excluded from block building (ibft.writeTransactions)
+	maxAccountDemotions = uint(10)
 )
 
 // errors
@@ -317,6 +321,9 @@ func (p *TxPool) Pop(tx *types.Transaction) {
 	// pop the top most promoted tx
 	account.promoted.pop()
 
+	//	successfully popping an account reset its demotions count to 0
+	account.demotions = 0
+
 	// update state
 	p.gauge.decrease(slotsRequired(tx))
 
@@ -370,6 +377,9 @@ func (p *TxPool) Drop(tx *types.Transaction) {
 	dropped = account.enqueued.clear()
 	clearAccountQueue(dropped)
 
+	//	reset the demotions counter
+	account.demotions = 0
+
 	p.eventManager.signalEvent(proto.EventType_DROPPED, tx.Hash)
 	p.logger.Debug("dropped account txs",
 		"num", droppedCount,
@@ -379,6 +389,15 @@ func (p *TxPool) Drop(tx *types.Transaction) {
 }
 
 func (p *TxPool) Demote(tx *types.Transaction) {
+	account := p.accounts.get(tx.From)
+	if account.demotions == maxAccountDemotions {
+		p.Drop(tx)
+
+		return
+	}
+
+	account.demotions++
+
 	p.eventManager.signalEvent(proto.EventType_DEMOTED, tx.Hash)
 }
 
@@ -684,6 +703,9 @@ func (p *TxPool) resetAccounts(stateNonces map[types.Address]uint64) {
 		//	append pruned
 		allPrunedPromoted = append(allPrunedPromoted, prunedPromoted...)
 		allPrunedEnqueued = append(allPrunedEnqueued, prunedEnqueued...)
+
+		//	new state for account -> demotions are reset to 0
+		account.demotions = 0
 	}
 
 	//	pool cleanup callback

--- a/txpool/txpool_test.go
+++ b/txpool/txpool_test.go
@@ -1030,6 +1030,7 @@ func TestPop(t *testing.T) {
 	assert.Equal(t, uint64(0), pool.gauge.read())
 	assert.Equal(t, uint64(0), pool.accounts.get(addr1).promoted.length())
 }
+
 func TestDrop(t *testing.T) {
 	pool, err := newTestPool()
 	assert.NoError(t, err)
@@ -1058,55 +1059,77 @@ func TestDrop(t *testing.T) {
 }
 
 func TestDemote(t *testing.T) {
-	//	create pool
-	pool, err := newTestPool()
-	assert.NoError(t, err)
-	pool.SetSigner(&mockSigner{})
+	t.Parallel()
 
-	//	send tx
-	go func() {
-		err := pool.addTx(local, newTx(addr1, 0, 1))
+	t.Run("Demote increments counter", func(t *testing.T) {
+		t.Parallel()
+		//	create pool
+		pool, err := newTestPool()
 		assert.NoError(t, err)
-	}()
-	go pool.handleEnqueueRequest(<-pool.enqueueReqCh)
-	pool.handlePromoteRequest(<-pool.promoteReqCh)
+		pool.SetSigner(&mockSigner{})
 
-	assert.Equal(t, uint64(1), pool.gauge.read())
-	assert.Equal(t, uint64(1), pool.accounts.get(addr1).getNonce())
-	assert.Equal(t, uint64(1), pool.accounts.get(addr1).promoted.length())
+		//	send tx
+		go func() {
+			err := pool.addTx(local, newTx(addr1, 0, 1))
+			assert.NoError(t, err)
+		}()
+		go pool.handleEnqueueRequest(<-pool.enqueueReqCh)
+		pool.handlePromoteRequest(<-pool.promoteReqCh)
 
-	//	call demote
-	pool.Prepare()
-	tx := pool.Peek()
-	pool.Demote(tx)
+		assert.Equal(t, uint64(1), pool.gauge.read())
+		assert.Equal(t, uint64(1), pool.accounts.get(addr1).getNonce())
+		assert.Equal(t, uint64(1), pool.accounts.get(addr1).promoted.length())
+		assert.Equal(t, uint(0), pool.accounts.get(addr1).demotions)
 
-	//	assert counter was incremented
-}
+		//	call demote
+		pool.Prepare()
+		tx := pool.Peek()
+		pool.Demote(tx)
 
-func TestDemoteToDrop(t *testing.T) {
-	//	create pool
-	pool, err := newTestPool()
-	assert.NoError(t, err)
-	pool.SetSigner(&mockSigner{})
+		assert.Equal(t, uint64(1), pool.gauge.read())
+		assert.Equal(t, uint64(1), pool.accounts.get(addr1).getNonce())
+		assert.Equal(t, uint64(1), pool.accounts.get(addr1).promoted.length())
 
-	//	set counter to max allowed skips
-	//	TODO
+		//	assert counter was incremented
+		assert.Equal(t, uint(1), pool.accounts.get(addr1).demotions)
+	})
 
-	//	send tx
-	go func() {
-		err := pool.addTx(local, newTx(addr1, 0, 1))
+	t.Run("Demote calls Drop", func(t *testing.T) {
+		t.Parallel()
+
+		//	create pool
+		pool, err := newTestPool()
 		assert.NoError(t, err)
-	}()
-	go pool.handleEnqueueRequest(<-pool.enqueueReqCh)
-	pool.handlePromoteRequest(<-pool.promoteReqCh)
+		pool.SetSigner(&mockSigner{})
 
-	//	call demote
-	pool.Prepare()
-	tx := pool.Peek()
-	pool.Demote(tx)
+		//	send tx
+		go func() {
+			err := pool.addTx(local, newTx(addr1, 0, 1))
+			assert.NoError(t, err)
+		}()
+		go pool.handleEnqueueRequest(<-pool.enqueueReqCh)
+		pool.handlePromoteRequest(<-pool.promoteReqCh)
 
-	//	assert: counter, account state
-	//	TODO
+		assert.Equal(t, uint64(1), pool.gauge.read())
+		assert.Equal(t, uint64(1), pool.accounts.get(addr1).getNonce())
+		assert.Equal(t, uint64(1), pool.accounts.get(addr1).promoted.length())
+
+		//	set counter to max allowed demotions
+		pool.accounts.get(addr1).demotions = maxAccountDemotions
+
+		//	call demote
+		pool.Prepare()
+		tx := pool.Peek()
+		pool.Demote(tx)
+
+		//	account was dropped
+		assert.Equal(t, uint64(0), pool.gauge.read())
+		assert.Equal(t, uint64(0), pool.accounts.get(addr1).getNonce())
+		assert.Equal(t, uint64(0), pool.accounts.get(addr1).promoted.length())
+
+		//	demotions are reset to 0
+		assert.Equal(t, uint(0), pool.accounts.get(addr1).demotions)
+	})
 }
 
 /* "Integrated" tests */

--- a/txpool/txpool_test.go
+++ b/txpool/txpool_test.go
@@ -1057,6 +1057,58 @@ func TestDrop(t *testing.T) {
 	assert.Equal(t, uint64(0), pool.accounts.get(addr1).promoted.length())
 }
 
+func TestDemote(t *testing.T) {
+	//	create pool
+	pool, err := newTestPool()
+	assert.NoError(t, err)
+	pool.SetSigner(&mockSigner{})
+
+	//	send tx
+	go func() {
+		err := pool.addTx(local, newTx(addr1, 0, 1))
+		assert.NoError(t, err)
+	}()
+	go pool.handleEnqueueRequest(<-pool.enqueueReqCh)
+	pool.handlePromoteRequest(<-pool.promoteReqCh)
+
+	assert.Equal(t, uint64(1), pool.gauge.read())
+	assert.Equal(t, uint64(1), pool.accounts.get(addr1).getNonce())
+	assert.Equal(t, uint64(1), pool.accounts.get(addr1).promoted.length())
+
+	//	call demote
+	pool.Prepare()
+	tx := pool.Peek()
+	pool.Demote(tx)
+
+	//	assert counter was incremented
+}
+
+func TestDemoteToDrop(t *testing.T) {
+	//	create pool
+	pool, err := newTestPool()
+	assert.NoError(t, err)
+	pool.SetSigner(&mockSigner{})
+
+	//	set counter to max allowed skips
+	//	TODO
+
+	//	send tx
+	go func() {
+		err := pool.addTx(local, newTx(addr1, 0, 1))
+		assert.NoError(t, err)
+	}()
+	go pool.handleEnqueueRequest(<-pool.enqueueReqCh)
+	pool.handlePromoteRequest(<-pool.promoteReqCh)
+
+	//	call demote
+	pool.Prepare()
+	tx := pool.Peek()
+	pool.Demote(tx)
+
+	//	assert: counter, account state
+	//	TODO
+}
+
 /* "Integrated" tests */
 
 // The following tests ensure that the pool's inner event loop

--- a/txpool/txpool_test.go
+++ b/txpool/txpool_test.go
@@ -612,7 +612,11 @@ func TestPromoteHandler(t *testing.T) {
 }
 
 func TestResetAccount(t *testing.T) {
+	t.Parallel()
+
 	t.Run("reset promoted", func(t *testing.T) {
+		t.Parallel()
+
 		testCases := []struct {
 			name     string
 			txs      []*types.Transaction
@@ -674,7 +678,10 @@ func TestResetAccount(t *testing.T) {
 			},
 		}
 		for _, test := range testCases {
+			test := test
 			t.Run(test.name, func(t *testing.T) {
+				t.Parallel()
+
 				pool, err := newTestPool()
 				assert.NoError(t, err)
 				pool.SetSigner(&mockSigner{})
@@ -725,6 +732,8 @@ func TestResetAccount(t *testing.T) {
 	})
 
 	t.Run("reset enqueued", func(t *testing.T) {
+		t.Parallel()
+
 		testCases := []struct {
 			name     string
 			txs      []*types.Transaction
@@ -807,7 +816,10 @@ func TestResetAccount(t *testing.T) {
 		}
 
 		for _, test := range testCases {
+			test := test
 			t.Run(test.name, func(t *testing.T) {
+				t.Parallel()
+
 				pool, err := newTestPool()
 				assert.NoError(t, err)
 				pool.SetSigner(&mockSigner{})
@@ -847,6 +859,8 @@ func TestResetAccount(t *testing.T) {
 	})
 
 	t.Run("reset enqueued and promoted", func(t *testing.T) {
+		t.Parallel()
+
 		testCases := []struct {
 			name     string
 			txs      []*types.Transaction
@@ -950,7 +964,10 @@ func TestResetAccount(t *testing.T) {
 		}
 
 		for _, test := range testCases {
+			test := test
 			t.Run(test.name, func(t *testing.T) {
+				t.Parallel()
+
 				pool, err := newTestPool()
 				assert.NoError(t, err)
 				pool.SetSigner(&mockSigner{})

